microservices/first/utils/db_manager.py

```diff
@@ -1,120 +1,142 @@
 import sqlite3
-from typing import Union
+from flask import g
+import json
 
 from . import deployment_status
 
 """
 Deployment Table Schema:
 id: Integer, Primary key identifying each row
-leader_ID: Integer, Indentifier of deployment
+leader_ID: TEXT, Indentifier of deployment
 repo_name: TEXT, Name of repo the deployment is working on
 repo_branch: TEXT, Branch of the repo being worked on
 status: TEXT, (DeploymentStatus in text form) Status of the deployment.
 
 Results Table Schema:
 id: Integer, Primary key identifying each row
-leader_ID: Integer, Foreign Key pointing to Deployment Table "leader_id". Deployment the result is associated with.
+deployment_ID: Integer, Foreign Key pointing to Deployment Table "id". Deployment the result is associated with.
 result: TEXT, one test result object.
 """
 class DBManager:
+
+  def _get_db(self) -> sqlite3.Connection:
+        db = getattr(g, '_database', None)
+        if db is None:
+            db = g._database = sqlite3.connect(self.db_file)
+        return db
+
+  def close_connection(self) -> None:
+      db = getattr(g, '_database', None)
+      if db is not None:
+          db.close()
+
   def __init__(self, db_file: str) -> None:
     """
     Initialize the database manager with the path to the database file.
     
     param db_file: The file path of the SQLite database.
     """
     self.db_file = db_file
-    self.conn = self.create_connection()
-
-  def create_connection(self) -> sqlite3.Connection:
-    """Create a database connection to the SQLite database."""
-    conn = sqlite3.connect(self.db_file)
-    return conn
-
-  def close_connection(self) -> None:
-    """Close the database connection."""
-    if self.conn:
-        self.conn.close()
 
   def create_deployments_table(self) -> None:
     """Create the deployments table in the database."""
-    c = self.conn.cursor()
+    c = self._get_db().cursor()
     c.execute(
               '''CREATE TABLE IF NOT EXISTS deployments (
               id INTEGER PRIMARY KEY, 
-              leader_ID INTEGER UNIQUE,
+              leader_ID TEXT,
               repo_name TEXT, 
               repo_branch TEXT,
               status TEXT DEFAULT 'STARTED')'''
             )
 
   def create_results_table(self) -> None:
     """Create the results table in the database."""
-    c = self.conn.cursor()
+    c = self._get_db().cursor()
     c.execute(
         '''CREATE TABLE IF NOT EXISTS results (
             id INTEGER PRIMARY KEY,
-            leader_ID INTEGER,
+            deployment_ID INTEGER,
             result TEXT,
-            FOREIGN KEY(leader_id) REFERENCES deployments(id)
+            FOREIGN KEY(deployment_ID) REFERENCES deployments(id)
         )'''
     )
+
+  def next_deployment_id(self) -> int:
```

Comment:
This should be made private

```suggestion
  def _next_deployment_id(self) -> int:
```